# ai03aTasks
# Machine Learning: Decision Trees
## Introduction & Setup

**Instructions:**
- Complete each task below by running the code cells
- Fill in the blanks and answer questions in markdown cells
- Save your work when finished
- Push this file to your GitHub "Machine Learning" Repo under the appropriate folder.

---
## Task 1: Import Libraries

Import the necessary libraries for machine learning with decision trees.  
Fill in the missing library names below.

In [3]:
import pandas as pd  # for working with data tables
from sklearn.model_selection import train_test_split  # Splits dataset into training and testing subsets
from sklearn.tree import DecisionTreeClassifier, plot_tree  # for building decision trees

import matplotlib.pyplot as plt  # Tfrom sklearn.metrics import accuracy_score, classification_report  # TODO: Fill in the two imports (accuracy_score, classification_report)ODO: Add comment explaining what this does

print("✓ Libraries imported successfully!")

✓ Libraries imported successfully!


---
## Task 2a: Load the Titanic Dataset
First go out and retreive the Titanic dataset from Kaggle using kagglehub

## Task 2b: Find the csv file in your .cache folder, then copy and paste it into the same folder your current environment is in (where this ipynb file is located!)

In [ ]:
import kagglehub
path = kagglehub.dataset_download("sakshisatre/titanic-dataset")
print("Dataset downloaded to:", path)


In [ ]:
# Now load the dataset from the local file (csv)
df = pd.read_csv("Titanic Dataset.csv")  # Make sure your csv 'Titanic Dataset.csv' file is in the right folder!

print("✓ Dataset loaded successfully!")
print(f"Number of rows: {len(df)}")
print(f"Number of columns: {len(df.columns)}")

---
## Task 3: Explore the Data

Use pandas methods to explore the dataset.

3a. Display the first 5 rows of the dataset

In [10]:
df = pd.read_csv("Titanic Dataset.csv")
df.head(5)
print("Dataset info:")
print(df.info())

Dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pclass     1309 non-null   int64  
 1   survived   1309 non-null   int64  
 2   name       1309 non-null   object 
 3   sex        1309 non-null   object 
 4   age        1046 non-null   float64
 5   sibsp      1309 non-null   int64  
 6   parch      1309 non-null   int64  
 7   ticket     1309 non-null   object 
 8   fare       1308 non-null   float64
 9   cabin      295 non-null    object 
 10  embarked   1307 non-null   object 
 11  boat       486 non-null    object 
 12  body       121 non-null    float64
 13  home.dest  745 non-null    object 
dtypes: float64(3), int64(4), object(7)
memory usage: 143.3+ KB
None


### 3b. Display the LAST 5 rows of the dataset

In [14]:
df = pd.read_csv("Titanic Dataset.csv")
df.tail()
print("Dataset info:")
print(df.info())


Dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pclass     1309 non-null   int64  
 1   survived   1309 non-null   int64  
 2   name       1309 non-null   object 
 3   sex        1309 non-null   object 
 4   age        1046 non-null   float64
 5   sibsp      1309 non-null   int64  
 6   parch      1309 non-null   int64  
 7   ticket     1309 non-null   object 
 8   fare       1308 non-null   float64
 9   cabin      295 non-null    object 
 10  embarked   1307 non-null   object 
 11  boat       486 non-null    object 
 12  body       121 non-null    float64
 13  home.dest  745 non-null    object 
dtypes: float64(3), int64(4), object(7)
memory usage: 143.3+ KB
None


### 3c. Get information about the dataset

In [16]:
import pandas as pd
df = pd.read_csv("Titanic Dataset.csv")
print("Dataset info:")
df.info()



Dataset info:
<class 'pandas.core.frame.DataFrame'>
RangeIndex: 1309 entries, 0 to 1308
Data columns (total 14 columns):
 #   Column     Non-Null Count  Dtype  
---  ------     --------------  -----  
 0   pclass     1309 non-null   int64  
 1   survived   1309 non-null   int64  
 2   name       1309 non-null   object 
 3   sex        1309 non-null   object 
 4   age        1046 non-null   float64
 5   sibsp      1309 non-null   int64  
 6   parch      1309 non-null   int64  
 7   ticket     1309 non-null   object 
 8   fare       1308 non-null   float64
 9   cabin      295 non-null    object 
 10  embarked   1307 non-null   object 
 11  boat       486 non-null    object 
 12  body       121 non-null    float64
 13  home.dest  745 non-null    object 
dtypes: float64(3), int64(4), object(7)
memory usage: 143.3+ KB


---
## Task 4: Analyze the Data

Answer these questions by running code and observing the output.

### 4a. How many total passengers are in the dataset?

In [17]:
# TODO: Write code to find the total number of rows
total_passengers = df.shape[0]
print(f"Total passengers in dataset: {total_passengers}")

Total passengers in dataset: 1309


### 4b. What are the column names (features)?

In [18]:
# TODO: Write code to display all column names
print("Column names:")
print(df.columns)  # Hint: use .columns

Column names:
Index(['pclass', 'survived', 'name', 'sex', 'age', 'sibsp', 'parch', 'ticket',
       'fare', 'cabin', 'embarked', 'boat', 'body', 'home.dest'],
      dtype='object')


### 4c. Display basic statistics for all numeric columns

In [19]:
print("Dataset statistics:")
print(df.describe())

Dataset statistics:
            pclass     survived          age        sibsp        parch  \
count  1309.000000  1309.000000  1046.000000  1309.000000  1309.000000   
mean      2.294882     0.381971    29.881138     0.498854     0.385027   
std       0.837836     0.486055    14.413493     1.041658     0.865560   
min       1.000000     0.000000     0.170000     0.000000     0.000000   
25%       2.000000     0.000000    21.000000     0.000000     0.000000   
50%       3.000000     0.000000    28.000000     0.000000     0.000000   
75%       3.000000     1.000000    39.000000     1.000000     0.000000   
max       3.000000     1.000000    80.000000     8.000000     9.000000   

              fare        body  
count  1308.000000  121.000000  
mean     33.295479  160.809917  
std      51.758668   97.696922  
min       0.000000    1.000000  
25%       7.895800   72.000000  
50%      14.454200  155.000000  
75%      31.275000  256.000000  
max     512.329200  328.000000  


---
## Task 5: Answer Questions About the Data

Use the output from your code above to answer these questions.

**Q1: How many passengers are in the dataset?**

A: 1309

**Q2: What is the average (mean) age of passengers?**

A:  29.881138

**Q3: What is the minimum fare paid? Maximum fare paid?**

A: Minimum: 0, Maximum: 512.329200

### Q4: How many people survived?

Hint: The `survived` column has 1 for survived, 0 for died

In [20]:
# TODO: Write code to count survivors
survivors = df['survived'].sum()  # sum() adds up all the 1s
print(f"Number of survivors: {survivors}")

Number of survivors: 500


**Answer:** 500

### Q5: What percentage of passengers survived?

In [21]:
# TODO: Calculate survival rate
survival_rate = (survivors / total_passengers) * 100
print(f"Survival rate: {survival_rate:.2f}%")

Survival rate: 38.20%


**Answer:** 38.20%

---
## Task 6: Check for Missing Values

Real-world data often has missing values. Let's find them!

In [22]:
# TODO: Write code to show how many missing values each column has
# howTo: Use df.isnull().sum()
import pandas as pd
missing_values = df.isnull().sum()
print("Missing values per column:")
print(missing_values)


Missing values per column:
pclass          0
survived        0
name            0
sex             0
age           263
sibsp           0
parch           0
ticket          0
fare            1
cabin        1014
embarked        2
boat          823
body         1188
home.dest     564
dtype: int64


**Which column has the most missing values?**

A: Body

---
## Challenge Tasks (Optional)

Try these additional challenges if you finish early!

### Challenge 1: What is the average age of survivors vs non-survivors?

In [ ]:
# TODO: Use df.groupby() to find average age for each survival status
# howTo: df.groupby('survived')['age'].mean()



### Challenge 2: How many males vs females were on the Titanic?

In [ ]:
# TODO: Use df['sex'].value_counts()


### Challenge 3: What was the survival rate for each passenger class?

In [ ]:
# TODO: Use df.groupby() to find survival rate for each class
# howTo: df.groupby('pclass')['survived'].mean()


---
## Reflection Questions

Answer these questions in your own words:

**1. Why is it important to inspect your data before building a model?**

Answer: So you can find errors.

**2. What does it mean if a column has missing values?**

Answer: there are rows missing in your dataset.

**3. Why might the Titanic dataset be useful for learning machine learning?**

Answer: its very small and easy to look through.

**4. What feature (column) do you think would be MOST important for predicting survival? Why?**

Answer: The name column. Someones name isnt going to change the survival prediction.

SyntaxError: invalid syntax (1231719442.py, line 1)